### ISSUE TITLE
How do I stub_responses for DynamoDB query method?

### ISSUE THREAD

**Issue by BilalBudhani** [aws/aws-sdk-ruby#822]
Hi there, 

I've been trying to stub_responses for below query

```ruby
count_query = dynamod_db_v2.query(table_name: 'Auth', index_name: 'Auth_UserId_Uid_Index', select: 'COUNT', key_conditions: {'userId' => {attribute_value_list: [@user.id], comparison_operator: 'EQ'}})
```

I've no clue how to stub it. I tried 

```ruby
db = Aws::DynamoDB::Client.new
db.stub_responses(:query, items: [], count: 1)
```
this doesn't work. I read #770 which says I've to provide original response from DynamoDB API but, I dunno how does it look like. Can you help with this?

**Comment by trevorrowe** [comment]
Its possible but there are few signifcant caveats currently:

* You must provide the protocol structure for DynamoDB which I'll demonstrate below. Its a bit awkward.
* Passing the structures expected by DynamoDB's protocol will bypass the initial issue, but it will also fail to return the simplified response values you probably expect.
* Because the currently implementation is a bug, when it is fixed to work as expected, this work-around will stop working.

With all of those disclaimer out of the way, the following example demonstrates stubbing all of the support types by DynamoDB. Note that the list, `:l` and map, `:m` entries may be recursive as they support mixed types and can be nested arbitrarily deep. Essentially every value is wrapped in an object that decorates its type.

```ruby
ddb = Aws::DynamoDB::Client.new(stub_responses:true)
ddb.stub_responses(:query, items: [
  {

    'string' => { s: 'value' }, 
    'string-set' => { ss: ['a', 'b', 'c'] },

    'number' => { n: '10.0' },
    'number-set' => { ns: ['10', '11', '12'] },

    'binary' => { b: 'data' },
    'binary-set' => { bs: ['data1', 'data2'] },

    'boolean' => { bool: true },
    'null' => { null: true },

    # lists, unlike sets allow mixed types
    'list' => { 
      l: [
        { s: 'value' },
        { n: '10' },
        { b: 'data' },
        { bool: false },
        { null: true },
        { l: [] },
        { m: {} },
      ]
    },

    # maps
    'map' => {
      m: {
        'string' => { s: 'value' },
        'number' => { n: '10' },
        'binary' => { b: 'data' },
        'boolean' => { bool: false },
        'null' => { null: true },
        'list' => { l: [] },
        'map' => { m: {} },
      }
    }
  }
], count: 1)
```

The response will not look like what you would expect because the simplified attributes plugin does not yet support stubbing. Here is what you could expect back. Notice the number attributes are always populated.

```ruby
ddb.query(table_name:"table-name")  
=> #<struct Aws::DynamoDB::Types::QueryOutput
 items=
  [{"string"=>#<struct Aws::DynamoDB::Types::AttributeValue s="value", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
    "string-set"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=["a", "b", "c"], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
    "number"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="10.0", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
    "number-set"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=["10", "11", "12"], bs=[], m={}, l=[], null=false, bool=false>,
    "binary"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b="data", ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
    "binary-set"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=["data1", "data2"], m={}, l=[], null=false, bool=false>,
    "boolean"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=true>,
    "null"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=true, bool=false>,
    "list"=>
     #<struct Aws::DynamoDB::Types::AttributeValue
      s="StringAttributeValue",
      n="NumberAttributeValue",
      b=nil,
      ss=[],
      ns=[],
      bs=[],
      m={},
      l=
       [#<struct Aws::DynamoDB::Types::AttributeValue s="value", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        #<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="10", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        #<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b="data", ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        #<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        #<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=true, bool=false>,
        #<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        #<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>],
      null=false,
      bool=false>,
    "map"=>
     #<struct Aws::DynamoDB::Types::AttributeValue
      s="StringAttributeValue",
      n="NumberAttributeValue",
      b=nil,
      ss=[],
      ns=[],
      bs=[],
      m=
       {"string"=>#<struct Aws::DynamoDB::Types::AttributeValue s="value", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        "number"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="10", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        "binary"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b="data", ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        "boolean"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        "null"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=true, bool=false>,
        "list"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>,
        "map"=>#<struct Aws::DynamoDB::Types::AttributeValue s="StringAttributeValue", n="NumberAttributeValue", b=nil, ss=[], ns=[], bs=[], m={}, l=[], null=false, bool=false>},
      l=[],
      null=false,
      bool=false>}],
 count=1,
 scanned_count=0,
 last_evaluated_key={},
 consumed_capacity=nil>
```

My personal recommendation is use DynamoDB local, where everything works as expected, or wait for this to be fixed in master. Im looking at a fix as part of the 2.1 release.

**Comment by BilalBudhani** [comment]
@trevorrowe Thanks for posting a brief explanation on this. I will try this out and update the issue with the conclusion. 

I would like to use DynamoDB local but, it just got launched few months back and lot of libraries / gems are still using older version of aws-sdk which doesn't support DynamoDB local. These constrain makes it really hard to migrate.

**Comment by trevorrowe** [comment]
I did open a follow up issue here: https://github.com/aws/aws-sdk-ruby/issues/823. Feel free to ask any questions you have, but the plan currently is to fix this for version 2.1.
